### PR TITLE
[SHELL32][SHLWAPI][BROWSEUI][EXPLORER] Save folder view state

### PR DESCRIPTION
**Pull request by whindsaks**
The final episode of the save Explorer window state trilogy. Saves/restores the Listview icon mode, columns and sort info per-folder. Store DefView state info in the registry using property bags provided by SHGetViewStatePropertyBag.

Notes:
- The Mode, FFlags, Sort and SortDir properties are Windows compatible.
- The ColInfo property and the legacy stream is not compatible because CFSFolder does not implement enough of IShellFolder2 to store the column GUIDs.
- I slightly adjusted the default CFSFolder column sizes so the name column is larger.
- Turning off the desktop icons with SSF_HIDEICONS/FWF_NOICONS does not actually work, this just deals with the save state, not the required HWND layout/ListView changes.
- Yes, Windows does save some changes when you bring up the logoff/shutdown dialog, not when you hit OK.

### CODE REVIEW

File: dll/win32/shell32/CDefView.cpp [hunk 1]
```diff
@@ -3069,11 +3120,190 @@ HRESULT WINAPI CDefView::AddPropertySheetPages(DWORD dwReserved, LPFNADDPROPSHEE
...
+    m_sortInfo.Direction = cvs.SortDir > 0 ? 1 : -1;
+    m_sortInfo.ListColumn = cvs.SortColId;
+    return hr;
+}
+
+HRESULT CDefView::SaveViewState(IStream *pStream)
+{
+    int sortcol = MapListColumnToFolderColumn(m_sortInfo.ListColumn);
+    PERSISTCLASSICVIEWSTATE cvs;
+    cvs.SortColId = sortcol >= 0 ? (WORD)sortcol : 0;
```
Comment (JoachimHenze) [inline dll/win32/shell32/CDefView.cpp:3228]: Why not simply like that? I thought that on several places. This looks much simpler to me:
```suggestion
    cvs.SortColId = (WORD)sortcol;
```
Do I miss anything?

Comment (whindsaks) [reply dll/win32/shell32/CDefView.cpp:3228]: Yes, it is negative on error.

File: dll/win32/shell32/CDefView.cpp [hunk 2]
```diff
@@ -3069,11 +3120,190 @@ HRESULT WINAPI CDefView::AddPropertySheetPages(DWORD dwReserved, LPFNADDPROPSHEE
...
+        UINT u32;
+        GetCurrentViewMode(&u32);
+        hr = SHPropertyBag_WriteDWORD(pPB, L"Mode", u32);
+        SHPropertyBag_WriteDWORD(pPB, L"FFlags", m_FolderSettings.fFlags);
+        SHPropertyBag_WriteDWORD(pPB, L"Sort", cvs.SortColId);
+        SHPropertyBag_WriteDWORD(pPB, L"SortDir", cvs.SortDir);
+        pStream = cols.Count ? SHCreateMemStream((LPBYTE)&cols, cbColumns) : NULL;
+        if (!pStream|| FAILED(SHPropertyBag_WriteStream(pPB, L"ColInfo", pStream)))
+            SHPropertyBag_Delete(pPB, L"ColInfo");
+#if 0 // TODO
```
Comment (JoachimHenze) [inline dll/win32/shell32/CDefView.cpp:3266]: **What** is supposed to be implemented here ultimately?

File: dll/win32/shell32/folders/CFSFolder.cpp [hunk 3]
```diff
@@ -528,11 +528,11 @@ CFSFolder::~CFSFolder()
 }
 
 static const shvheader GenericSFHeader[] = {
-    {IDS_SHV_COLUMN_NAME, SHCOLSTATE_TYPE_STR | SHCOLSTATE_ONBYDEFAULT, LVCFMT_LEFT, 15},
+    {IDS_SHV_COLUMN_NAME, SHCOLSTATE_TYPE_STR | SHCOLSTATE_ONBYDEFAULT, LVCFMT_LEFT, 18},
```
Comment (JoachimHenze) [inline dll/win32/shell32/folders/CFSFolder.cpp:531]: What was the magic 15? What is the magic 18?

File: dll/win32/shell32/CDefView.cpp (hunk 2 above)
Comment (whindsaks) [reply dll/win32/shell32/CDefView.cpp:3266]: The listview item positions when auto-arrange is off.

File: dll/win32/shell32/folders/CFSFolder.cpp (hunk 3 above)
Comment (whindsaks) [reply dll/win32/shell32/folders/CFSFolder.cpp:531]: The default width of the column.

File: dll/win32/shlwapi/propbag.cpp [hunk 4]
```diff
@@ -1463,7 +1463,7 @@ CViewStatePropertyBag::_GetMRUSlots(
         return hr;
 
     hr = pMruList->QueryPidl(pidl, cSlots, puSlots, pcSlots);
-    if (hr == S_OK && MODE_CAN_WRITE(dwMode))
+    if (hr == S_OK || MODE_CAN_WRITE(dwMode))
```
Comment (JoachimHenze) [inline dll/win32/shlwapi/propbag.cpp:1466]: Why that change? Is this a bugfix that would fix anything outside of your PR? If not, why is it needed for your work?

File: dll/win32/shell32/CDefView.cpp (hunk 2 above)
Comment (JoachimHenze) [reply dll/win32/shell32/CDefView.cpp:3266]:
```suggestion
#if 0 // TODO: Implement listview item positions when auto-arrange is off
```

File: base/shell/explorer/traywnd.cpp [hunk 5]
```diff
@@ -660,8 +660,18 @@ class CTrayWindow :
         return S_OK;
     }
 
+    void SaveState()
+    {
+        if (SHRestricted(REST_NOSAVESET))
+            return ;
```
Comment (JoachimHenze) [inline base/shell/explorer/traywnd.cpp:666]: 666 - the number of the beast:
```suggestion
            return;
```

File: dll/win32/shell32/CDefView.cpp [hunk 6]
```diff
@@ -1215,11 +1265,12 @@ BOOL CDefView::_Sort(int Col)
...
     hColumn.mask = HDI_FORMAT;
     Header_GetItem(hHeader, m_sortInfo.ListColumn, &hColumn);
-    hColumn.fmt &= (m_sortInfo.bIsAscending ? ~HDF_SORTDOWN : ~HDF_SORTUP );
-    hColumn.fmt |= (m_sortInfo.bIsAscending ?  HDF_SORTUP   : HDF_SORTDOWN);
+    hColumn.fmt &= ~(HDF_SORTDOWN | HDF_SORTUP);
+    hColumn.fmt |= (m_sortInfo.Direction > 0 ?  HDF_SORTUP : HDF_SORTDOWN);
     Header_SetItem(hHeader, m_sortInfo.ListColumn, &hColumn);
 
     /* Sort the list, using the current values of ListColumn and bIsAscending */
+    ASSERT(m_sortInfo.Direction == 1 || m_sortInfo.Direction == -1);
```
Comment (JoachimHenze) [inline dll/win32/shell32/CDefView.cpp:1273]: Why can't the direction be == 0 here?
(See line 1250 for when it is set to 0.)

File: dll/win32/shlwapi/propbag.cpp (hunk 4 above)
Comment (whindsaks) [reply dll/win32/shlwapi/propbag.cpp:1466]: This is a bugfix (and the other one further down fixes a crash).  It will not fix anything outside this PR because SHGetViewStatePropertyBag only exists for DefView to get a property storage interface it can use like it does in this PR.

Without this change, nothing will ever be saved. The old code is `if (FoundExistingEntry && CanWrite)` and since nobody has saved before, this will never be true.

I admit this looks somewhat weird. It would be better if we could keep the && and instead check with SUCCEEDED. I will investigate this again to see if CMruPidlList::QueryPidl can return S_FALSE in the case we care about.

Comment (JoachimHenze) [reply dll/win32/shlwapi/propbag.cpp:1466]: Thank you for the explanation. I do consider this discussion to be resolved to my satisfaction, but I don't want to collapse your reply.

File: dll/win32/shell32/CDefView.cpp (hunk 6 above)
Comment (whindsaks) [reply dll/win32/shell32/CDefView.cpp:1273]: And on line 1253 it makes sure it is not 0 before doing the actual sort.

**Review by JoachimHenze — APPROVED** [review]
Approved, (aside the nitpick on line666). Please wait a week to allow some more experienced devs having a look at it as well. Thx for working on that.

File: dll/win32/shlwapi/propbag.cpp (hunk 4 above)
Comment (whindsaks) [reply dll/win32/shlwapi/propbag.cpp:1466]: It is possible the bug is somewhere else, @katahiromz might know since I believe he wrote this. Something has to be done, otherwise SHGetViewStatePropertyBag will never allow you to write anything.

File: sdk/include/reactos/shlwapi_undoc.h [hunk 7]
```diff
@@ -187,7 +187,6 @@ HRESULT WINAPI SHPropertyBag_WritePOINTS(IPropertyBag *ppb, LPCWSTR pszPropName,
 HRESULT WINAPI SHPropertyBag_WriteRECTL(IPropertyBag *ppb, LPCWSTR pszPropName, const RECTL *prcl);
 
 HRESULT WINAPI SHCreatePropertyBagOnMemory(_In_ DWORD dwMode, _In_ REFIID riid, _Out_ void **ppvObj);
-
```
Comment (HBelusca) [inline sdk/include/reactos/shlwapi_undoc.h:190]: Please revert, this doesn't help with readability.

File: sdk/include/reactos/shlwapi_undoc.h [hunk 8]
```diff
@@ -203,6 +202,12 @@ SHCreatePropertyBagOnProfileSection(
     _In_ DWORD dwMode,
     _In_ REFIID riid,
     _Out_ void **ppvObj);
+EXTERN_C HRESULT WINAPI
```
Comment (HBelusca) [inline sdk/include/reactos/shlwapi_undoc.h:205]:
```suggestion

EXTERN_C HRESULT WINAPI
```
(same reason as above.)

File: dll/win32/shell32/shelldesktop/CDesktopBrowser.cpp [hunk 9]
```diff
@@ -228,10 +235,13 @@ HRESULT CDesktopBrowser::Initialize(IShellDesktopTray *ShellDesk)
...
         return hRet;
 
+    SHELLSTATE shellstate;
+    shellstate.fHideIcons = FALSE;
+    SHGetSetSettings(&shellstate, SSF_HIDEICONS, FALSE);
     FOLDERSETTINGS fs;
     RECT rcShellView = {0,0,0,0};
     fs.ViewMode = FVM_ICON;
-    fs.fFlags = FWF_DESKTOP | FWF_NOCLIENTEDGE | FWF_NOSCROLL | FWF_TRANSPARENT;
+    fs.fFlags = FWF_DESKTOP | FWF_NOCLIENTEDGE | FWF_NOSCROLL | FWF_TRANSPARENT | FWF_AUTOARRANGE | (shellstate.fHideIcons ? FWF_NOICONS : 0);
```
Comment (HBelusca) [inline dll/win32/shell32/shelldesktop/CDesktopBrowser.cpp:244]:
```suggestion
    fs.fFlags = FWF_DESKTOP | FWF_NOCLIENTEDGE | FWF_NOSCROLL | FWF_TRANSPARENT |
                FWF_AUTOARRANGE | (shellstate.fHideIcons ? FWF_NOICONS : 0);
```

File: dll/win32/shell32/shelldesktop/CDesktopBrowser.cpp [hunk 10]
```diff
@@ -349,9 +359,15 @@ HRESULT STDMETHODCALLTYPE CDesktopBrowser::SetToolbarItems(LPTBBUTTON lpButtons,
     return E_NOTIMPL;
 }
 
+HRESULT STDMETHODCALLTYPE CDesktopBrowser::GetPropertyBag(long flags, REFIID riid, void **ppv)
+{
+    DWORD deskpidl = 0;
+    return SHGetViewStatePropertyBag((LPITEMIDLIST)&deskpidl, L"Desktop", flags | SHGVSPB_ROAM, riid, ppv);
```
Comment (HBelusca) [inline dll/win32/shell32/shelldesktop/CDesktopBrowser.cpp:365]:
```suggestion
    ITEMIDLIST deskpidl = {0};
    return SHGetViewStatePropertyBag(&deskpidl, L"Desktop", flags | SHGVSPB_ROAM, riid, ppv);
```
to avoid the ugly cast.

File: dll/win32/shell32/wine/shellord.c [hunk 11]
```diff
@@ -1822,20 +1823,24 @@ BOOL WINAPI ReadCabinetState(CABINETSTATE *cs, int length)
 	if ( (r != ERROR_SUCCESS) || (cs->cLength < sizeof(*cs)) ||
 		(cs->cLength != length) )
 	{
+		SHELLSTATE shellstate;
+		shellstate.fWin95Classic = FALSE;
+		SHGetSetSettings(&shellstate, SSF_WIN95CLASSIC, FALSE);
+
 		TRACE("Initializing shell cabinet settings\n");
-		memset(cs, 0, sizeof(*cs));
```
Comment (HBelusca) [inline dll/win32/shell32/wine/shellord.c:1826]: why removing it?

File: dll/win32/shlwapi/propbag.cpp [hunk 12]
```diff
@@ -1917,13 +1917,9 @@ SHGetViewStatePropertyBag(
...
     }
-
-    g_pCachedBag.Attach(pBag);
-
-    hr = g_pCachedBag->QueryInterface(riid, ppv);
-
+    g_pCachedBag = pBag;
     ::LeaveCriticalSection(&g_csBagCacheLock);
-    return hr;
+    return pBag->QueryInterface(riid, ppv);
```
Comment (HBelusca) [inline dll/win32/shlwapi/propbag.cpp:1922]: why the change?

File: dll/win32/shell32/CDefView.cpp [hunk 13]
```diff
@@ -60,8 +59,18 @@ enum  {
...
+    INT8    Direction;
+    bool    bLoadedFromViewState;
+    bool    bColumnIsFolderColumn;
+    UINT8   Reserved; // Unused
     INT     ListColumn;
+
+    enum { UNSPECIFIEDCOLUMN = -1 };
+    void Reset()
+    {
+        *(UINT*)this = 0;
```
Comment (HBelusca) [inline dll/win32/shell32/CDefView.cpp:71]: Which fields above this will set to zero?

File: dll/win32/shell32/CDefView.cpp [hunk 14]
```diff
@@ -1197,11 +1245,13 @@ BOOL CDefView::_Sort(int Col)
...
         if (prevCol == Col)
-            m_sortInfo.bIsAscending = !m_sortInfo.bIsAscending;
+            m_sortInfo.Direction *= -1;
         else
-            m_sortInfo.bIsAscending = TRUE;
+            m_sortInfo.Direction = 0;
         m_sortInfo.ListColumn = Col;
     }
+    if (!m_sortInfo.Direction)
+        m_sortInfo.Direction += 1;
```
Comment (HBelusca) [inline dll/win32/shell32/CDefView.cpp:1254]: So these two lines, just actually set `m_sortInfo.Direction` to 1 ? (since this is in the case `!m_sortInfo.Direction` being TRUE, i.e. `m_sortInfo.Direction == 0`).
If so, why not just doing `...Direction = 1;` instead of the += 1 ?

And further so, why not setting Direction = 1 in the else-case that is above (line 1250)?

File: dll/win32/shell32/CDefView.cpp [hunk 15]
```diff
@@ -1453,9 +1504,9 @@ HRESULT CDefView::FillList(BOOL IsRefreshCommand)
 
     /* sort the array */
     int sortCol = -1;
-    if (!IsRefreshCommand) // Are we loading for the first time?
+    if (!IsRefreshCommand && !m_sortInfo.bLoadedFromViewState) // Are we loading for the first time?
     {
-        m_sortInfo.bIsAscending = TRUE;
+        m_sortInfo.Direction = 0;
```
Comment (HBelusca) [inline dll/win32/shell32/CDefView.cpp:1509]: What's the meaning of Direction == 0 ? (Maybe this would provide me with some extra info regarding my previous question)

File: dll/win32/shell32/CDefView.cpp [hunk 16]
```diff
@@ -3069,11 +3120,190 @@ HRESULT WINAPI CDefView::AddPropertySheetPages(DWORD dwReserved, LPFNADDPROPSHEE
     return S_OK;
 }
 
+static HRESULT Read(IStream *pS, LPVOID buffer, ULONG cb)
+{
+    ULONG read;
+    HRESULT hr = pS->Read(buffer, cb, &read);
+    return FAILED(hr) ? hr : (cb == read ? S_OK : HResultFromWin32(ERROR_MORE_DATA));
```
Comment (HBelusca) [inline dll/win32/shell32/CDefView.cpp:3127]: Why do you need this extra HRESULT conversion?
What about [`IStream_Read() aka. SHIStream_Read()`](https://git.reactos.org/?p=reactos.git;a=blob;f=dll/win32/shlwapi/istream.c;hb=957bb89548ba90484b67c553559b92405eb8e39c#l541) ?

File: dll/win32/shell32/CDefView.cpp [hunk 17]
```diff
@@ -3069,11 +3120,190 @@ HRESULT WINAPI CDefView::AddPropertySheetPages(DWORD dwReserved, LPFNADDPROPSHEE
...
+    m_FolderSettings.fFlags &= ~cvs.VALIDFWF;
+    m_FolderSettings.fFlags |= cvs.FolderSettings.fFlags & cvs.VALIDFWF;
+    if (SUCCEEDED(hrColumns))
+    {
+        BOOL failed = FALSE;
+        if ((m_LoadColumnsList = DPA_Create(cols.Count)) != NULL)
+        {
+            for (UINT i = 0; i < cols.Count; ++i)
+            {
+                failed |= !DPA_SetPtr(m_LoadColumnsList, i, (void*)(SIZE_T)cols.Columns[i]);
```
Comment (HBelusca) [inline dll/win32/shell32/CDefView.cpp:3208]:
```suggestion
                failed |= !DPA_SetPtr(m_LoadColumnsList, i, UlongToPtr(cols.Columns[i]));
```

File: dll/win32/shell32/CDefView.cpp [hunk 18]
```diff
@@ -3069,11 +3120,190 @@ HRESULT WINAPI CDefView::AddPropertySheetPages(DWORD dwReserved, LPFNADDPROPSHEE
...
+    if (SUCCEEDED(hr))
+    {
+        UINT u32;
+        GetCurrentViewMode(&u32);
+        hr = SHPropertyBag_WriteDWORD(pPB, L"Mode", u32);
+        SHPropertyBag_WriteDWORD(pPB, L"FFlags", m_FolderSettings.fFlags);
+        SHPropertyBag_WriteDWORD(pPB, L"Sort", cvs.SortColId);
+        SHPropertyBag_WriteDWORD(pPB, L"SortDir", cvs.SortDir);
+        pStream = cols.Count ? SHCreateMemStream((LPBYTE)&cols, cbColumns) : NULL;
+        if (!pStream|| FAILED(SHPropertyBag_WriteStream(pPB, L"ColInfo", pStream)))
```
Comment (HBelusca) [inline dll/win32/shell32/CDefView.cpp:3264]:
```suggestion
        if (!pStream || FAILED(SHPropertyBag_WriteStream(pPB, L"ColInfo", pStream)))
```

File: dll/win32/shell32/CDefView.cpp [hunk 19]
```diff
@@ -3069,11 +3120,190 @@ HRESULT WINAPI CDefView::AddPropertySheetPages(DWORD dwReserved, LPFNADDPROPSHEE
...
+    HRESULT hr = S_OK;
+    if (pStream)
+    {
+        pStream->AddRef();
+        goto stream;
+    }
+    hr = IUnknown_QueryServicePropertyBag(m_pShellBrowser, SHGVSPB_FOLDER, IID_PPV_ARG(IPropertyBag, &pPB));
+    if (SUCCEEDED(hr))
+    {
+        UINT u32;
```
Comment (HBelusca) [inline dll/win32/shell32/CDefView.cpp:3257]: uViewMode, or uView ...

File: dll/win32/shell32/CDefView.cpp [hunk 20]
```diff
@@ -3069,11 +3120,190 @@ HRESULT WINAPI CDefView::AddPropertySheetPages(DWORD dwReserved, LPFNADDPROPSHEE
...
 HRESULT WINAPI CDefView::SaveViewState()
 {
-    FIXME("(%p) stub\n", this);
+    if (!(m_FolderSettings.fFlags & FWF_NOBROWSERVIEWSTATE))
+        return SaveViewState(NULL);
+    return S_FALSE;
+}
 
-    return S_OK;
+#define UPDATEFOLDERVIEWFLAGS(bits, bit, on) ( (bits) = ((bits) & ~(bit)) | ((on) ? (bit) : 0) )
```
Comment (HBelusca) [inline dll/win32/shell32/CDefView.cpp:3301]: "on" -> "enable" / "set"

File: dll/win32/shell32/CDefView.cpp [hunk 21]
```diff
@@ -3456,6 +3703,7 @@ HRESULT STDMETHODCALLTYPE CDefView::CreateViewWindow3(IShellBrowser *psb, IShell
         else
             FIXME("Ignoring unrecognized VID %s\n", debugstr_guid(view_id));
     }
+    const UINT requestedviewmode = m_FolderSettings.ViewMode;
```
Comment (HBelusca) [inline dll/win32/shell32/CDefView.cpp:3706]: requestedViewMode

File: dll/win32/shell32/CDefView.cpp [hunk 22]
```diff
@@ -3182,6 +3412,18 @@ HRESULT WINAPI CDefView::GetItemObject(UINT uItem, REFIID riid, LPVOID *ppvOut)
...
 
+FOLDERVIEWMODE CDefView::GetDefaultViewMode()
+{
+    FOLDERVIEWMODE mode = ((m_FolderSettings.fFlags & FWF_DESKTOP) || !IsOS(OS_SERVERADMINUI)) ? FVM_ICON : FVM_DETAILS;
+    FOLDERVIEWMODE temp;
+    if (SUCCEEDED(_DoFolderViewCB(SFVM_DEFVIEWMODE, 0, (LPARAM)&temp)))
+    {
+        if (temp >= FVM_FIRST && temp <= FVM_LAST)
+            mode = temp;
+    }
```
Comment (HBelusca) [inline dll/win32/shell32/CDefView.cpp:3423]: Nested if-s may be collapsed
```suggestion
    if (SUCCEEDED(_DoFolderViewCB(SFVM_DEFVIEWMODE, 0, (LPARAM)&temp)) &&
        (temp >= FVM_FIRST && temp <= FVM_LAST))
    {
        mode = temp;
    }
```

File: dll/win32/shell32/CDefView.cpp [hunk 23]
```diff
@@ -3069,11 +3120,190 @@ HRESULT WINAPI CDefView::AddPropertySheetPages(DWORD dwReserved, LPFNADDPROPSHEE
...
+    hr = IUnknown_QueryServicePropertyBag(m_pShellBrowser, SHGVSPB_FOLDER, IID_PPV_ARG(IPropertyBag, &pPB));
+    if (SUCCEEDED(hr))
+    {
+        UINT u32;
+        GetCurrentViewMode(&u32);
+        hr = SHPropertyBag_WriteDWORD(pPB, L"Mode", u32);
+        SHPropertyBag_WriteDWORD(pPB, L"FFlags", m_FolderSettings.fFlags);
+        SHPropertyBag_WriteDWORD(pPB, L"Sort", cvs.SortColId);
+        SHPropertyBag_WriteDWORD(pPB, L"SortDir", cvs.SortDir);
+        pStream = cols.Count ? SHCreateMemStream((LPBYTE)&cols, cbColumns) : NULL;
```
Comment (HBelusca) [inline dll/win32/shell32/CDefView.cpp:3263]: Just to confirm: the SHCreateMemStream adds an initial reference on the pStream?

File: dll/win32/shell32/wine/shellord.c (hunk 11 above)
Comment (whindsaks) [reply dll/win32/shell32/wine/shellord.c:1826]: Because there was only one member that was not being set. We are setting the defaults here and should be specific about everything.

File: dll/win32/shell32/CDefView.cpp [hunk 24]
```diff
@@ -71,6 +80,25 @@ typedef struct
 // to call TrackPopupMenu and let it use the 0 value as an indication that the menu was canceled
 #define CONTEXT_MENU_BASE_ID 1
 
+struct PERSISTCOLUMNS
```
Comment (cbialorucki) [inline dll/win32/shell32/CDefView.cpp:83]: Could you explain why we need signatures for these structs? Apologies if this is a stupid question.

File: dll/win32/shell32/CDefView.cpp (hunk 23 above)
Comment (whindsaks) [reply dll/win32/shell32/CDefView.cpp:3263]: Yes

File: dll/win32/shlwapi/propbag.cpp (hunk 12 above)
Comment (whindsaks) [reply dll/win32/shlwapi/propbag.cpp:1922]: Because we hold the lock for fewer instructions. Changing `Attach` to `=` fixed a crash and acting on the plain pBag interface outside the lock is safe.

File: dll/win32/shell32/CDefView.cpp (hunk 13 above)
Comment (whindsaks) [reply dll/win32/shell32/CDefView.cpp:71]: All of them except ListColumn.

File: dll/win32/shell32/CDefView.cpp (hunk 1 above)
Comment (cbialorucki) [reply dll/win32/shell32/CDefView.cpp:3228]: You could do `max((WORD)sortcol, 0)`.

File: dll/win32/shell32/CDefView.cpp (hunk 14 above)
Comment (whindsaks) [reply dll/win32/shell32/CDefView.cpp:1254]: I can't set it on line 1250 because that path is not always hit. In general, the data in m_sortInfo is not fully trusted because several things outside DefView gets to play with the sorting.

File: dll/win32/shell32/CDefView.cpp (hunk 15 above)
Comment (whindsaks) [reply dll/win32/shell32/CDefView.cpp:1509]: You can think of it as undefined with a slight lean towards sort ascending. If nobody outside DefView has any opinion on the matter either, the sort function in your previous question just has to pick a direction and it will pick ascending.

File: dll/win32/shell32/CDefView.cpp (hunk 16 above)
Comment (whindsaks) [reply dll/win32/shell32/CDefView.cpp:3127]: I can't call the helper function in shlwapi for whatever reason. Probably related to wine code being C. The correct name to use outside shlwapi would be IStream_Read but wine calls it SHIStream_Read because C has those COM helper defines, one of them being IStream_Read.

File: dll/win32/shell32/CDefView.cpp (hunk 24 above)
Comment (whindsaks) [reply dll/win32/shell32/CDefView.cpp:83]: Need is a strong word. We are reading binary data from the registry here and in this case, just an array of UINTs. Without a signature, I have no way of verifying that this data is at least somewhat correct.

Failing to read the data just means we forget to restore the previously configures columns in this folder and just end up with the default column configuration instead,

File: dll/win32/shell32/CDefView.cpp (hunk 1 above)
Comment (whindsaks) [reply dll/win32/shell32/CDefView.cpp:3228]: > You could do `max((WORD)sortcol, 0)`.

The WORD cast in this case would mean 0xffff for -1 and that is larger than 0?

Besides, I don't think it makes the code clearer. I prefer: `valid ? value : default`.

File: dll/win32/browseui/shellbrowser.cpp [hunk 25]
```diff
@@ -406,7 +407,7 @@ class CShellBrowser :
     // *** IServiceProvider methods ***
     STDMETHOD(QueryService)(REFGUID guidService, REFIID riid, void **ppvObject) override;
 
-    // *** IShellBowserService methods ***
```
[the remark this answers is not in the document]
Comment (whindsaks) [reply dll/win32/browseui/shellbrowser.cpp:409]: It's meee, Mario!

File: dll/win32/shell32/CDefView.cpp (hunk 1 above)
Comment (cbialorucki) [reply dll/win32/shell32/CDefView.cpp:3228]: `max` is defined as 
`#define MAX(a,b) ((a) > (b) ? (a) : (b))`
So no, it wouldn't be evaluated as 'larger'.
I do think using `max` is cleaner, but it isn't a big issue.

Comment (cbialorucki) [reply dll/win32/shell32/CDefView.cpp:3228]: Fair enough. I forgot for a second that WORD is typedef'd as an unsigned.

**Review by cbialorucki — APPROVED** [review]

File: dll/win32/shlwapi/propbag.cpp (hunk 4 above)
Comment (katahiromz) [reply dll/win32/shlwapi/propbag.cpp:1466]: [bag.patch](https://github.com/user-attachments/files/16170493/bag.patch)

Can you try this patch? Thanks.

File: dll/win32/shell32/CDefView.cpp (hunk 24 above)
Comment (cbialorucki) [reply dll/win32/shell32/CDefView.cpp:83]: Good enough for me, though I do think when we get a transactional registry this will be unnecessary.

File: dll/win32/shlwapi/propbag.cpp (hunk 4 above)
Comment (whindsaks) [reply dll/win32/shlwapi/propbag.cpp:1466]: > Can you try this patch?

If I undo my fix and apply this instead it stops working.

Comment (katahiromz) [reply dll/win32/shlwapi/propbag.cpp:1466]: Don't undo. Please use both yours and mine.

File: dll/win32/browseui/shellbrowser.cpp (hunk 25 above)
[the remark this answers is not in the document]
Comment (HBelusca) [reply dll/win32/browseui/shellbrowser.cpp:409]: Reminds me of the famous ["bowser" driver](https://learn.microsoft.com/en-us/archive/blogs/larryosterman/fun-with-names) :joy:

File: dll/win32/shlwapi/propbag.cpp (hunk 12 above)
Comment (HBelusca) [reply dll/win32/shlwapi/propbag.cpp:1922]: > and acting on the plain pBag interface outside the lock is safe.

But that pBag interface is now also what's now stored in the global `g_pCachedBag`, no?
So what would happen if in the meantime (just after the LeaveCriticalSection call and before the pBag->QueryInterface call), we get separate threads that call to _any_ of the other functions in dll/win32/shlwapi/propbag.cpp (those that acquire `g_csBagCacheLock` because they are going to act/read on the `g_pCachedBag`)

File: dll/win32/shlwapi/propbag.cpp (hunk 4 above)
Comment (whindsaks) [reply dll/win32/shlwapi/propbag.cpp:1466]: Why? It already works without your patch.

Comment (katahiromz) [reply dll/win32/shlwapi/propbag.cpp:1466]: Please add a comment as `"FIXME: HACK!"`.

File: dll/win32/shlwapi/propbag.cpp (hunk 12 above)
Comment (whindsaks) [reply dll/win32/shlwapi/propbag.cpp:1922]: There is no in-between LeaveCriticalSection and XYZ to worry about. The reference count of pBag is >= 1. CBasePropertyBag uses interlocked functions to handle the reference count...

File: dll/win32/shell32/CDefView.cpp (hunk 16 above)
Comment (HBelusca) [reply dll/win32/shell32/CDefView.cpp:3127]: > The correct name to use outside shlwapi would be IStream_Read [...]

we are outside of shlwapi here, so maybe it would work? (since that IStream_Read function is exported).... unless it would use here the autogenerated macros from the headers? Dunno.

File: dll/win32/shlwapi/propbag.cpp (hunk 12 above)
Comment (HBelusca) [reply dll/win32/shlwapi/propbag.cpp:1922]: @katahiromz Any thoughts about this? (you wrote the code)

File: dll/win32/shell32/CDefView.cpp [hunk 26]
```diff
@@ -60,8 +59,18 @@ enum  {
 
 typedef struct
 {
-    BOOL    bIsAscending;
+    INT8    Direction;
+    bool    bLoadedFromViewState;
+    bool    bColumnIsFolderColumn;
+    UINT8   Reserved; // Unused
```
Comment (HBelusca) [inline dll/win32/shell32/CDefView.cpp:65]: I guess the member name already conveys the information ^^

File: dll/win32/shell32/CDefView.cpp [hunk 27]
```diff
@@ -757,20 +793,21 @@ BOOL CDefView::CreateList()
...
-    ListExStyle = 0;
+              LVS_SHAREIMAGELISTS | LVS_EDITLABELS | LVS_AUTOARRANGE; // FIXME: Remove LVS_AUTOARRANGE when the view is able to save ItemPos
+    dwExStyle = (m_FolderSettings.fFlags & FWF_NOCLIENTEDGE) ? 0 : WS_EX_CLIENTEDGE;
+    ListExStyle = LVS_EX_INFOTIP | LVS_EX_LABELTIP;
 
     if (m_FolderSettings.fFlags & FWF_DESKTOP)
     {
         m_FolderSettings.fFlags |= FWF_NOCLIENTEDGE | FWF_NOSCROLL;
         dwStyle |= LVS_ALIGNLEFT;
+        // LVS_EX_REGIONAL?
```
Comment (HBelusca) [inline dll/win32/shell32/CDefView.cpp:804]: It seems this style is somewhat deprecated (and would work only on LVS_ICON styled listviews).
Does Windows XP/2003 set it?

File: dll/win32/shell32/wine/shellord.c (hunk 11 above)
Comment (whindsaks) [reply dll/win32/shell32/wine/shellord.c:1826]: This struct has a version member so undefined padding is not a big deal **and** this struct has not changed since NT4 **and** it will take another 8 new bools to end up with hitting the padding...but I will zero it out for you...

File: dll/win32/shell32/CDefView.cpp (hunk 27 above)
Comment (whindsaks) [reply dll/win32/shell32/CDefView.cpp:804]: I think it's used on the desktop listview, perhaps only on multimon...

**Review by HBelusca — APPROVED** [review]
